File: iOS/AUIKitCore/Sources/Service/Impl/AUIMicSeatServiceImpl.swift
//
//  AUIMicSeatServiceImpl.swift
//  Pods
//
//  Created by wushengtao on 2023/2/21.
//


import Foundation
import AgoraRtmKit

private let kSeatAttrKey = "micSeat"

private enum AUIMicSeatCmd: String {
    case leaveSeatCmd = "leaveSeatCmd"
    case enterSeatCmd = "enterSeatCmd"
    case kickSeatCmd = "kickSeatCmd"
    case muteAudioCmd = "muteAudioCmd"
    case closeSeatCmd = "closeSeatCmd"
}

//麦位Service实现(纯端上修改KV)
@objc open class AUIMicSeatServiceImpl: NSObject {
    private var respDelegates: NSHashTable<AUIMicSeatRespDelegate> = NSHashTable<AUIMicSeatRespDelegate>.weakObjects()
    private var channelName: String!
    private let rtmManager: AUIRtmManager!
    
    private var micSeats:[Int: AUIMicSeatInfo] = [:]
    
    private var mapCollection: AUIMapCollection!
        
    deinit {
        aui_info("deinit AUIMicSeatServiceImpl", tag: "AUIMicSeatServiceImpl")
    }
    
    public init(channelName: String, rtmManager: AUIRtmManager) {
        self.rtmManager = rtmManager
        self.channelName = channelName
        self.mapCollection = AUIMapCollection(channelName: channelName, observeKey: kSeatAttrKey, rtmManager: rtmManager)
        super.init()
        mapCollection.subscribeWillMerge {[weak self] publisherId, dataCmd, updateMap, currentMap in
            return self?.metadataWillMerge(publiserId: publisherId, dataCmd: dataCmd, updateMap: updateMap, currentMap: currentMap)
        }
        mapCollection.subscribeAttributesDidChanged {[weak self] channelName, key, value in
            self?.onAttributesDidChanged(channelName: channelName, key: key, value: value)
        }
        aui_info("init AUIMicSeatServiceImpl", tag: "AUIMicSeatServiceImpl")
    }
}

extension AUIMicSeatServiceImpl: AUIMicSeatServiceDelegate {
    
    public func getRoomContext() -> AUIRoomContext {
        return AUIRoomContext.shared
    }
    
    public func getChannelName() -> String {
        return channelName
    }
    
    public func bindRespDelegate(delegate: AUIMicSeatRespDelegate) {
        respDelegates.add(delegate)
    }
    
    public func unbindRespDelegate(delegate: AUIMicSeatRespDelegate) {
        respDelegates.remove(delegate)
    }
    
    public func enterSeat(seatIndex: Int, callback: @escaping (NSError?) -> ()) {
        let user = AUIUserThumbnailInfo()
        user.userId = getRoomContext().currentUserInfo.userId
        user.userAvatar = getRoomContext().currentUserInfo.userAvatar
        user.userName = getRoomContext().currentUserInfo.userName
        let value = [
            "\(seatIndex)": [
                "owner": user.yy_modelToJSONObject(),
                "micSeatStatus": AUILockSeatStatus.user.rawValue
            ]
        ]
        self.mapCollection.mergeMetaData(valueCmd: AUIMicSeatCmd.enterSeatCmd.rawValue,
                                         value: value,
                                         filter: nil,
                                         callback: callback)
    }
    
    public func leaveSeat(callback: @escaping (NSError?) -> ()) {
        leaveSeat(userId: getRoomContext().currentUserInfo.userId, callback: callback)
    }
    
    public func pickSeat(seatIndex: Int, user: AUIUserThumbnailInfo, callback: @escaping (NSError?) -> ()) {
        let value = [
            "\(seatIndex)": [
                "owner": user.yy_modelToJSONObject(),
                "micSeatStatus": AUILockSeatStatus.user.rawValue
            ]
        ]
        self.mapCollection.mergeMetaData(valueCmd: AUIMicSeatCmd.enterSeatCmd.rawValue,
                                         value: value,
                                         filter: nil,
                                         callback: callback)
    }
    
    public func kickSeat(seatIndex: Int, callback: @escaping (NSError?) -> ()) {
        let value = [
            "\(seatIndex)": [
                "owner": AUIUserThumbnailInfo().yy_modelToJSONObject(),
                "micSeatStatus": AUILockSeatStatus.idle.rawValue
            ]
        ]
        self.mapCollection.mergeMetaData(valueCmd: AUIMicSeatCmd.kickSeatCmd.rawValue,
                                         value: value,
                                         filter: nil,
                                         callback: callback)
    }
    
    public func muteAudioSeat(seatIndex: Int, isMute: Bool, callback: @escaping (NSError?) -> ()) {
        let value = [
            "\(seatIndex)": [
                "isMuteAudio": isMute
            ]
        ]
        self.mapCollection.mergeMetaData(valueCmd: AUIMicSeatCmd.muteAudioCmd.rawValue,
                                         value: value,
                                         filter: nil,
                                         callback: callback)
    }
    
    public func muteVideoSeat(seatIndex: Int, isMute: Bool, callback: @escaping AUICallback) {
    }
    
    public func closeSeat(seatIndex: Int, isClose: Bool, callback: @escaping (NSError?) -> ()) {
        var value: [String: Any] = [:]
        self.micSeats.forEach { (k: Int, v: AUIMicSeatInfo) in
            if k == seatIndex {
                var status: AUILockSeatStatus = .idle
                if isClose {
                    status = .locked
                } else if v.user?.isEmpty() ?? true {
                    status = .idle
                } else {
                    status = .user
                }
                value["\(seatIndex)"] = [
                    "micSeatStatus": status.rawValue
                ]
            }
        }
        
        //TODO: value.isEmpty
        
        self.mapCollection.mergeMetaData(valueCmd: AUIMicSeatCmd.closeSeatCmd.rawValue,
                                         value: value,
                                         filter: nil,
                                         callback: callback)
    }
    
    public func isOnMicSeat(userId: String) -> Bool {
        for (_, seat) in micSeats {
            if seat.user?.userId == userId {
                return true
            }
        }
        
        return false
    }
    
    public func getMicSeatIndex(userId: String) -> Int {
        for (idx, seat) in micSeats {
            if seat.user?.userId == userId {
                return idx
            }
        }
        return -1
    }
}

//MARK: set metadata
extension AUIMicSeatServiceImpl {
    private func leaveSeat(userId: String, callback: @escaping (NSError?) -> ()) {
        var value: [String: Any] = [:]
        self.micSeats.forEach { (k: Int, v: AUIMicSeatInfo) in
            if userId == v.user?.userId {
                value["\(k)"] = [
                    "owner": AUIUserThumbnailInfo().yy_modelToJSONObject(),
                    "micSeatStatus": AUILockSeatStatus.idle.rawValue
                ]
            }
        }
        
        //TODO: value.isEmpty
        
        self.mapCollection.mergeMetaData(valueCmd: AUIMicSeatCmd.leaveSeatCmd.rawValue,
                                         value: value,
                                         filter: nil,
                                         callback: callback)
    }
    
    private func onAttributesDidChanged(channelName: String, key: String, value: AUIAttributesModel) {
        if key == kSeatAttrKey {
            aui_info("recv seat attr did changed \(value)", tag: "AUIMicSeatServiceImpl")
            guard let map = value.getMap() as? [String: [String: Any]] else {return}
            map.values.forEach { element in
                guard let micSeat = AUIMicSeatInfo.yy_model(with: element) else {return}
                aui_info(" micSeat.islock \(micSeat.lockSeat) micSeat.Index = \(micSeat.seatIndex)", tag: "AUIMicSeatServiceImpl")
                let index: Int = Int(micSeat.seatIndex)
                let origMicSeat = self.micSeats[index]
                
                self.micSeats[index] = micSeat
                self.respDelegates.allObjects.forEach { delegate in
                    if let origUser = origMicSeat?.user, origUser.userId.count > 0, micSeat.user?.userId ?? "" != origUser.userId {
                        delegate.onAnchorLeaveSeat(seatIndex: index, user: origUser)
                    }
                    
                    if let user = micSeat.user, user.userId.count > 0, origMicSeat?.user?.userId ?? "" != user.userId {
                        delegate.onAnchorEnterSeat(seatIndex: index, user: user)
                    }
                    
                    if origMicSeat?.lockSeat ?? .idle != micSeat.lockSeat {
                        delegate.onSeatClose(seatIndex: index, isClose: micSeat.lockSeat == .locked)
                    }
                    
                    if origMicSeat?.muteAudio ?? false != micSeat.muteAudio {
                        delegate.onSeatAudioMute(seatIndex: index, isMute: micSeat.muteAudio)
                    }
                    
                    if origMicSeat?.muteVideo ?? false != micSeat.muteVideo {
                        delegate.onSeatVideoMute(seatIndex: index, isMute: micSeat.muteVideo)
                    }
                    /*
                    if origMicSeat?.muteVideo != micSeat.muteVideo {
                        delegate.onSeatVideoMute(seatIndex: index, isMute: micSeat.muteVideo)
                    }
                    
                    if origMicSeat?.muteAudio != micSeat.muteAudio {
                        delegate.onSeatAudioMute(seatIndex: index, isMute: micSeat.muteAudio)
                    }
                     */
                }
            }
        }
    }
    
    private func metadataWillMerge(publiserId: String,
                                   dataCmd: String?,
                                   updateMap: [String: Any],
                                   currentMap: [String: Any]) -> NSError? {
        guard let dataCmd = AUIMicSeatCmd(rawValue: dataCmd ?? ""),
              updateMap.keys.count == 1,
              let seatIndex = Int(updateMap.keys.first ?? ""),
              let value = updateMap["\(seatIndex)"] else {
            return AUICommonError.unknown.toNSError()
        }
        
        let owner = (value as? [String: Any])?["owner"] as? [String: Any]
        var userId: String = owner?["userId"] as? String ?? ""
        switch dataCmd {
        case .enterSeatCmd:
            if self.micSeats.values.contains(where: { $0.user?.userId == userId }) {
                return AUICommonError.micSeatAlreadyEnter.toNSError()
            }
            guard let seat = self.micSeats[seatIndex],
                  seat.lockSeat == .idle,
                  seat.user?.isEmpty() ?? true else {
                return AUICommonError.micSeatNotIdle.toNSError()
            }
            break
        case .leaveSeatCmd:
            if seatIndex == 0 {
                return AUICommonError.noPermission.toNSError()
            }
            guard micSeats[seatIndex]?.user?.userId == publiserId
                    || getRoomContext().isRoomOwner(channelName: channelName, userId: publiserId) else {
                return AUICommonError.userNoEnterSeat.toNSError()
            }
            var err: NSError?
            //TODO: onSeatWillLeave不需要metaData？
            let metaData = NSMutableDictionary()//rtmLeaveSeatMetaData(userId: userId)
            userId = self.micSeats[seatIndex]?.user?.userId ?? ""
            for obj in respDelegates.allObjects {
                err = obj.onSeatWillLeave?(userId: userId, metaData: metaData)
                if let err = err {
                    return err
                }
            }
            break
        case .kickSeatCmd:
            if seatIndex == 0 {
                return AUICommonError.noPermission.toNSError()
            }
            var err: NSError? = nil
            let metaData = NSMutableDictionary()
            userId = self.micSeats[seatIndex]?.user?.userId ?? ""
            for obj in respDelegates.allObjects {
                err = obj.onSeatWillLeave?(userId: userId, metaData: metaData)
                if let _ = err {
                    return err
                }
            }
            break
        case .muteAudioCmd:
            break
        case .closeSeatCmd:
            break
        }
        
        return nil
    }
}

extension AUIMicSeatServiceImpl {
    public func initService(completion: @escaping ((NSError?) -> ())){
        guard let roomInfo = getRoomContext().roomInfoMap[channelName] else {
            completion(AUICommonError.unknown.toNSError())
            return
        }
        var seatMap: [String: [String: Any]] = [:]
        for i in 0...roomInfo.micSeatCount {
            let seat = AUIMicSeatInfo()
            seat.seatIndex = i
            if i == 0 {
                seat.user = getRoomContext().currentUserInfo
                seat.lockSeat = .user
            }
            seatMap["\(i)"] = seat.yy_modelToJSONObject() as? [String : Any]
        }
        
        let data = try! JSONSerialization.data(withJSONObject: seatMap, options: .prettyPrinted)
        let str = String(data: data, encoding: .utf8)!
        var metaData = [String: String]()
        metaData[kSeatAttrKey] = str
        rtmManager.setBatchMetadata(channelName: channelName,
                                    lockName: "",
                                    metadata: metaData,
                                    completion: completion)
    }
    
    public func cleanUserInfo(userId: String, completion: @escaping ((NSError?) -> ())) {
//        let micSeatMetaData = rtmLeaveSeatMetaData(userId: userId)
//        let str = micSeatMetaData.yy_modelToJSONString() ?? ""
//        var metaData = [String: String]()
//        metaData[kSeatAttrKry] = str
//        
//        rtmManager.setBatchMetadata(channelName: channelName,
//                                    lockName: kRTM_Referee_LockName,
//                                    metadata: metaData,
//                                    completion: completion)
        
        leaveSeat(userId: userId, callback: completion)
    }
    
    public func deinitService(completion:  @escaping  ((NSError?) -> ())) {
//        rtmManager.cleanBatchMetadata(channelName: channelName,
//                                      lockName: kRTM_Referee_LockName,
//                                      removeKeys: [kSeatAttrKry],
//                                      completion: completion)
        mapCollection.cleanMetaData(callback: completion)
    }
}

File: Android/auikit-service/src/main/java/io/agora/auikit/service/imp/AUIMicSeatServiceImpl.kt
package io.agora.auikit.service.imp

import android.util.Log
import io.agora.auikit.model.AUIMicSeatInfo
import io.agora.auikit.model.AUIMicSeatStatus
import io.agora.auikit.model.AUIRoomContext
import io.agora.auikit.model.AUIUserThumbnailInfo
import io.agora.auikit.service.IAUIMicSeatService
import io.agora.auikit.service.callback.AUICallback
import io.agora.auikit.service.callback.AUIException
import io.agora.auikit.service.collection.AUIAttributesModel
import io.agora.auikit.service.collection.AUICollectionException
import io.agora.auikit.service.collection.AUIMapCollection
import io.agora.auikit.service.rtm.AUIRtmManager
import io.agora.auikit.utils.AUILogger
import io.agora.auikit.utils.GsonTools
import io.agora.auikit.utils.ObservableHelper

const val kSeatAttrKey = "micSeat"

enum class AUIMicSeatCmd {
    initSeatCmd,
    leaveSeatCmd,
    enterSeatCmd,
    kickSeatCmd,
    muteAudioCmd,
    closeSeatCmd
}

class AUIMicSeatServiceImpl(
    private val channelName: String,
    private val rtmManager: AUIRtmManager
) : IAUIMicSeatService {

    private val observableHelper =
        ObservableHelper<IAUIMicSeatService.AUIMicSeatRespObserver>()

    private var micSeats = mutableMapOf<Int, AUIMicSeatInfo>()

    private val mapCollection = AUIMapCollection(channelName, kSeatAttrKey, rtmManager)


    init {
        mapCollection.subscribeWillMerge(this::metadataWillMerge)
        mapCollection.subscribeAttributesDidChanged(this::onAttributeChanged)
    }

    override fun initService(completion: AUICallback?) {
        if (!AUIRoomContext.shared().isRoomOwner(channelName)) {
            return
        }
        val roomInfo = roomContext.getRoomInfo(channelName) ?: return
        val seatMap = mutableMapOf<String, Any>()
        for (i in 0 until roomInfo.micSeatCount) {
            val seat = AUIMicSeatInfo()
            seat.seatIndex = i
            if (i == 0) {
                seat.user = roomContext.currentUserInfo
                seat.seatStatus = AUIMicSeatStatus.used
            }else{
                seat.user = AUIUserThumbnailInfo()
            }
            seatMap.put(i.toString(), seat)
        }
        AUILogger.logger().d(
            "AUIMicSeatServiceImp",
            "initService >> currentUid=${roomContext.currentUserInfo.userId} arbiterUid=${
                roomContext.getArbiter(
                    channelName
                )?.lockOwnerId()
            }"
        )
        mapCollection.updateMetaData(
            AUIMicSeatCmd.initSeatCmd.name,
            seatMap,
        ) {
            completion?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }

    override fun deInitService(completion: AUICallback?) {
        super.deInitService(completion)
        mapCollection.cleanMetaData {
            completion?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
        mapCollection.release()
    }

    override fun registerRespObserver(observer: IAUIMicSeatService.AUIMicSeatRespObserver?) {
        observableHelper.subscribeEvent(observer)
    }

    override fun unRegisterRespObserver(observer: IAUIMicSeatService.AUIMicSeatRespObserver?) {
        observableHelper.unSubscribeEvent(observer)
    }

    override fun enterSeat(seatIndex: Int, callback: AUICallback?) {
        mapCollection.mergeMetaData(
            valueCmd = AUIMicSeatCmd.enterSeatCmd.name,
            value = mapOf(
                Pair(
                    seatIndex.toString(),
                    mapOf(
                        Pair("owner", GsonTools.beanToMap(roomContext.currentUserInfo)),
                        Pair("micSeatStatus", AUIMicSeatStatus.used)
                    )
                )
            )
        ) {
            callback?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }

    override fun leaveSeat(callback: AUICallback?) {
        val userId = roomContext.currentUserInfo.userId
        val micSeat = micSeats.values.find { it.user?.userId == userId }

        if (micSeat == null) {
            callback?.onResult(
                AUIException(
                    AUIException.ERROR_CODE_SEAT_NOT_ENTER,
                    "user not on seat"
                )
            )
            return
        }
        micSeat.seatIndex

        mapCollection.mergeMetaData(
            valueCmd = AUIMicSeatCmd.leaveSeatCmd.name,
            value = mapOf(
                Pair(
                    micSeat.seatIndex.toString(),
                    mapOf(
                        Pair("owner", GsonTools.beanToMap(AUIUserThumbnailInfo())),
                        Pair("micSeatStatus", AUIMicSeatStatus.idle)
                    )
                )
            )
        ) {
            callback?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }

    override fun autoEnterSeat(callback: AUICallback?) {
        var toIndex: Int? = null
        val sortedKeys = micSeats.keys.sortedBy { key -> key }
        for (key in sortedKeys) {
            val value = micSeats[key]
            if (value?.seatStatus == AUIMicSeatStatus.idle) {
                toIndex = key
                break
            }
        }
        if (toIndex != null) {
            enterSeat(toIndex, callback)
        } else {
            callback?.onResult(
                AUIException(
                    -1,
                    "can not find empty mic seat"
                )
            )
        }
    }

    override fun kickSeat(seatIndex: Int, callback: AUICallback?) {
        if (!micSeats.containsKey(seatIndex)) {
            callback?.onResult(AUIException(-1, "The $seatIndex seat is not exist."))
            return
        }
        mapCollection.mergeMetaData(
            valueCmd = AUIMicSeatCmd.kickSeatCmd.name,
            value = mapOf(
                Pair(
                    seatIndex.toString(),
                    mapOf(
                        Pair("owner", GsonTools.beanToMap(AUIUserThumbnailInfo())),
                        Pair("micSeatStatus", AUIMicSeatStatus.idle)
                    )
                )
            )
        ) {
            callback?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }


    override fun pickSeat(seatIndex: Int, userId: String, callback: AUICallback?) {
        // do nothing
        throw RuntimeException("Not implement yet.")
    }


    override fun muteAudioSeat(seatIndex: Int, isMute: Boolean, callback: AUICallback?) {
        if (!micSeats.containsKey(seatIndex)) {
            callback?.onResult(AUIException(-1, "The $seatIndex seat is not exist."))
            return
        }
        mapCollection.mergeMetaData(
            valueCmd = AUIMicSeatCmd.muteAudioCmd.name,
            value = mapOf(Pair(seatIndex.toString(), mapOf(Pair("isMuteAudio", isMute))))
        ) {
            callback?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }

    override fun muteVideoSeat(seatIndex: Int, isMute: Boolean, callback: AUICallback?) {
        // do nothing
        throw RuntimeException("Not implement yet.")
    }

    override fun closeSeat(seatIndex: Int, isClose: Boolean, callback: AUICallback?) {
        if (!micSeats.containsKey(seatIndex)) {
            callback?.onResult(AUIException(-1, "The $seatIndex seat is not exist."))
            return
        }
        val micSeat = micSeats[seatIndex]

        var status = AUIMicSeatStatus.idle
        if (isClose) {
            status = AUIMicSeatStatus.locked
        } else if (micSeat?.user?.userId?.isNotEmpty() == true) {
            status = AUIMicSeatStatus.used
        }
        mapCollection.mergeMetaData(
            valueCmd = AUIMicSeatCmd.closeSeatCmd.name,
            value = mapOf(Pair(seatIndex.toString(), mapOf(Pair("micSeatStatus", status))))
        ) {
            callback?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }

    override fun onClickInvited(index: Int) {
        observableHelper.notifyEventHandlers {
            it.onShowInvited(index)
        }
    }

    override fun getMicSeatInfo(seatIndex: Int): AUIMicSeatInfo? {
        return micSeats[seatIndex]
    }

    override fun getMicSeatIndex(userId: String): Int {
        var index = -1
        micSeats.forEach { (key, value) ->
            if (value.user?.userId == userId) {
                index = key
            }
        }
        return index
    }

    override fun getMicSeatSize(): Int {
        return micSeats.size
    }

    override fun getChannelName() = channelName

    private fun onAttributeChanged(channelName: String, key: String, value: AUIAttributesModel) {
        if (key != kSeatAttrKey) {
            return
        }
        Log.d("mic_seat_update", "class: ${value.javaClass}")
        val map: Map<String, Any> = HashMap()
        val seats = value.getMap() ?: GsonTools.toBean(GsonTools.beanToString(value), map.javaClass)
        Log.d("mic_seat_update", "seats: $seats")
        seats?.values?.forEach {
            val newSeatInfo =
                GsonTools.toBean(GsonTools.beanToString(it), AUIMicSeatInfo::class.java) ?: return
            val index = newSeatInfo.seatIndex
            val oldSeatInfo = micSeats[index]
            micSeats[index] = newSeatInfo
            val newSeatUserId = newSeatInfo.user?.userId ?: ""
            val oldSeatUserId = oldSeatInfo?.user?.userId ?: ""
            if (oldSeatUserId.isEmpty() && newSeatUserId.isNotEmpty()) {
                Log.d("mic_seat_update", "onAnchorEnterSeat: $it")
                val newUser = newSeatInfo.user ?: return
                observableHelper.notifyEventHandlers { delegate ->
                    delegate.onAnchorEnterSeat(index, newUser)
                }
            }
            if (oldSeatUserId.isNotEmpty() && newSeatUserId.isEmpty()) {
                Log.d("mic_seat_update", "onAnchorLeaveSeat: $it")
                val originUser = oldSeatInfo?.user ?: return
                observableHelper.notifyEventHandlers { delegate ->
                    delegate.onAnchorLeaveSeat(index, originUser)
                }
            }
            if ((oldSeatInfo?.seatStatus ?: AUIMicSeatStatus.idle) != newSeatInfo.seatStatus &&
                (oldSeatInfo?.seatStatus == AUIMicSeatStatus.locked || newSeatInfo.seatStatus == AUIMicSeatStatus.locked)
            ) {
                Log.d("mic_seat_update", "onSeatClose: $it")
                observableHelper.notifyEventHandlers { delegate ->
                    delegate.onSeatClose(index, (newSeatInfo.seatStatus == AUIMicSeatStatus.locked))
                }
            }
            if (oldSeatInfo?.muteAudio != newSeatInfo.muteAudio) {
                Log.d("mic_seat_update", "onSeatAudioMute: $it")
                observableHelper.notifyEventHandlers { delegate ->
                    delegate.onSeatAudioMute(index, newSeatInfo.muteAudio)
                }
            }
            if (oldSeatInfo?.muteVideo != newSeatInfo.muteVideo) {
                Log.d("mic_seat_update", "onSeatVideoMute: $it")
                observableHelper.notifyEventHandlers { delegate ->
                    delegate.onSeatVideoMute(index, newSeatInfo.muteVideo)
                }
            }
        }
    }

    private fun metadataWillMerge(
        publisherId: String,
        valueCmd: String?,
        newValue: Map<String, Any>,
        oldValue: Map<String, Any>
    ): AUICollectionException? {
        val seatInfoPair = newValue.toList()[0]
        val seatIndex = seatInfoPair.first.toInt()
        val seatInfoMap = seatInfoPair.second as? Map<*, *>
        var userId = (seatInfoMap?.get("owner") as? Map<*, *>)?.get("userId") as? String
        when (valueCmd) {
            AUIMicSeatCmd.enterSeatCmd.name -> {
                if (micSeats.values.any { it.user?.userId == userId }) {
                    return AUICollectionException.ErrorCode.unknown.toException("code: ${AUIException.ERROR_CODE_SEAT_ALREADY_ENTER}")
                }
                val seatInfo = micSeats[seatIndex]
                userId = seatInfo?.user?.userId ?: ""
                if (seatInfo?.user != null && seatInfo.seatStatus != AUIMicSeatStatus.idle) {
                    return AUICollectionException.ErrorCode.unknown.toException(
                        "${
                            AUIException(
                                AUIException.ERROR_CODE_SEAT_NOT_IDLE,
                                "mic seat not idle"
                            )
                        }"
                    )
                }
            }

            AUIMicSeatCmd.leaveSeatCmd.name -> {
                if (seatIndex == 0) {
                    return AUICollectionException.ErrorCode.unknown.toException(
                        "${
                            AUIException(
                                AUIException.ERROR_CODE_PERMISSION_LEAK,
                                ""
                            )
                        }"
                    )
                }
                if (micSeats[seatIndex]?.user?.userId != publisherId || roomContext.isRoomOwner(
                        channelName,
                        publisherId
                    )
                ) {
                    return AUICollectionException.ErrorCode.unknown.toException(
                        "${
                            AUIException(
                                AUIException.ERROR_CODE_SEAT_NOT_ENTER,
                                ""
                            )
                        }"
                    )
                }
                userId = micSeats[seatIndex]?.user?.userId ?: ""
                val metadata = mutableMapOf<String, String>()
                var error: AUIException? = null
                observableHelper.notifyEventHandlers {
                    error = it.onSeatWillLeave(userId, metadata)?.let { return@notifyEventHandlers }
                }
                return if (error == null) null else AUICollectionException.ErrorCode.unknown.toException(
                    "$error"
                )
            }

            AUIMicSeatCmd.kickSeatCmd.name -> {
                if (seatIndex == 0) {
                    return AUICollectionException.ErrorCode.unknown.toException("${AUIException(AUIException.ERROR_CODE_PERMISSION_LEAK, "")}")
                }
                userId = micSeats[seatIndex]?.user?.userId ?: ""
                val metadata = mutableMapOf<String, String>()
                var error: AUIException? = null
                observableHelper.notifyEventHandlers {
                    error = it.onSeatWillLeave(userId, metadata)?.let { return@notifyEventHandlers }
                }
                return if (error == null) null else AUICollectionException.ErrorCode.unknown.toException(
                    "$error"
                )
            }
        }
        return null
    }
}